### ISSUE TITLE
Build the logit-parity quality gate for perf changes

### ISSUE THREAD

**Issue by spokvulcan** [spokvulcan/tesseract#233]
Child of map #230.

## Question

(Task, AFK.) The map's standing rule: no perf change lands without proof of zero output-quality regression. Build the automated gate once, so every optimization ticket can cite it:

- Harness mode (e.g. `--logit-parity-gate`) that runs a fixed prompt battery (short + long-context, text-only; a vision case if cheap) through the **current** stack and a **baseline** (recorded golden logits, or an A/B of two builds), comparing final-position logits and greedy continuations.
- Tolerance model: bitwise where kernel shapes are unchanged; where a change legitimately alters chunk shapes, bound by the measured FP-noise floor (the `chunkShapeNoiseControl` precedent in `ParoQuantVLMSmokeRunner` — argmax must still agree, maxAbsDiff within the recorded floor).
- Golden storage: per-model, per-context recorded logits checked into `research/` or regenerated on demand from a pinned baseline build — decide which is less brittle.
- Must run on Release (Debug MLX is ~20× slower and not the shipping numeric path).
- Existing ParoQuantTests (incl. the CPU-reference parity test) stay as the vendor-level layer; this gate covers the app-level end-to-end path.

Resolution = the gate exists, documented in the map so optimization tickets reference it as their acceptance bar.

**Comment by spokvulcan** [comment]
## Resolution — the gate exists, and it was validated against controls

`--logit-parity-record` / `--logit-parity-gate` (`LogitParityGateRunner.swift`), wired into `scripts/dev.sh logit-parity-record|logit-parity-gate`. **Release-only**, as the ticket required. Full write-up: `research/perf-230/logit-parity-gate.md`.

Every optimization ticket on this map now cites this as its acceptance bar:

```bash
# baseline build
scripts/dev.sh logit-parity-record --bench-model-id qwen3.6-35b-a3b-paro \
    --bench-contexts 128,8192,32768 --parity-max-new 64 --bench-kv-bits 8
# candidate build   (exit 0 = pass, exit 1 = regression)
scripts/dev.sh logit-parity-gate --bench-model-id qwen3.6-35b-a3b-paro --bench-kv-bits 8
```

### Two design decisions the measurements forced

**1. Decode is teacher-forced.** Free-running greedy decode is chaotic. Measured on the 35B PARO MoE at 32K under kv8, merely changing the prefill chunk size 1024→768 — which cannot alter the mathematics, only FP summation shapes — diverged the greedy continuation at **token 29 of 64**. An untethered comparison measures chaos, not quality. The gate feeds the *golden* token at each step, so every step sees an identical prefix. Under teacher forcing the same benign change yields 0/64 prediction changes (fp16) and 1/64 (kv8).

**2. Two logit vectors, independent tolerances.** The final-position prefill logits are structurally **blind to `kvBits`**: `quantizedKVStart` defaults to 0 and `maybeQuantizeKVCache` (`KVCache.swift:1859`) converts the cache only *after* the step-0 forward. An early version compared only that vector — a deliberate negative control (kv8 build vs fp16-KV goldens, a genuine numeric change) reported `maxAbsDiff = 0.0` and **passed**. The gate now also captures the **last decode step**'s logits, which see the quantized cache; the same control now reports `prefill=0.0, decode=1.189` and fails. Each vector carries its own floor — folding them into one `max()` let a large prefill floor mask a real decode regression.

### Tolerance model

- **bitwise** (default): both maxAbsDiffs ≤ the recorded run-to-run repeatability floor, zero teacher-forced prediction changes. That floor measures **0.0** at every context and both KV configs — this stack is exactly reproducible — so bitwise means bitwise.
- **floor** (`--parity-allow-floor`), for changes that legitimately alter kernel/chunk shapes. All three required: prefill argmax agrees; each vector within its own recorded chunk-shape FP-noise floor (the `chunkShapeNoiseControl` precedent generalized — each case re-run at an alternate prefill step during recording); and the candidate changes **no more predictions than that known-benign shape change did**.

### Validation (35B PARO MoE, Release, quiet machine)

| control | required | observed |
|:--|:--|:--|
| C1 unchanged kv8 build vs kv8 goldens | PASS bitwise, exit 0 | PASS bitwise, `maxAbsDiff=0.0`, 0/64 mismatches, all 3 contexts |
| C2 kv8 build vs fp16-KV goldens, **with** `--parity-allow-floor` | FAIL, exit 1 | FAIL — caught at 129 (`decode=1.189 > 0.420`) and 31,885 (1/64 predictions changed vs control 0/64) |
| C3 fp16 build @ prefill step 768 + `--parity-allow-floor` | PASS floor, exit 0 | PASS floor, 0/64 mismatches |
| C4 same, **without** the floor flag | refuse, exit 1 | exit 1 |

C2 is the load-bearing one: floor mode does not whitewash a real numeric change. Honest limit — in C2 the 7,981-token case passed floor (`decode=0.375` vs floor `0.420`), so per-case sensitivity is not uniform. The gate caught it at 2 of 3 contexts and failed overall. A single context is not a gate; the battery is.

### Side finding for the kvBits question (feeds #231)

**kv8 amplifies numeric sensitivity.** At 32K its decode-vector chunk-shape noise floor is **1.166 vs 0.287** for fp16 (4×), and it is the only config where a benign chunk-size change flips a prediction (1/64 vs 0/64). Independent of the 9–13% decode cost #217 measured, kv8 is also the numerically fragile option.

### Golden storage

Regenerated on demand from a baseline build, not checked in — they encode this machine's GPU and the exact model files. Stored under the app sandbox at `…/benchmark/logit-parity-goldens/<modelID>/<kv-config>/`. `--bench-contexts` selects what to *record*; gate mode always re-runs the whole recorded battery.

### Layering + maintenance

`ParoQuantTests` (incl. the CPU-reference parity test) stay the vendor-level numeric layer; this gate covers the app-level end-to-end path. Vision cases are out of the battery (`ParoQuantVLMSmokeRunner` covers VLM numerics). `captureCase` replicates `TokenIterator.step` semantics — **any change to the vendor iterator's step semantics must be mirrored there**, or the gate stops measuring the production path. That coupling is called out in the type doc.

Code is in the working tree (uncommitted, pending owner review): `LogitParityGateRunner.swift`, `TesseractApp.swift`, `ParoParityBenchRunner.swift` (prompt builder shared), `scripts/dev.sh`.
